### PR TITLE
Added support for temporal segmentation data in encoder decoder factory

### PR DESCRIPTION
**Pull request by rijuld**

### CODE REVIEW

**Comment by Joao-L-S-Almeida** [comment]
Hi, @rijuld Maybe you prefer to convert your PR to a draft and add the prefix [WIP] (work in progree) while works on it. 
When you feel it's ready to be merged, you can convert it back to PR and add me as reviewer.

**Comment by Joao-L-S-Almeida** [comment]
Thank you for your contribution.

File: terratorch/models/encoder_decoder_factory.py [hunk 1]
```diff
@@ -20,6 +21,31 @@
...
+            self.pooling = torch.mean
+        elif pooling == "max":
+            self.pooling = torch.max
+        else:
+            msg = "Pooling must be 'mean' or 'max'"
+            raise ValueError(msg)
+
+    def forward(self, x):
+        # x is a list of tensors, each corresponding to a different timestamp
+        features = [self.encoder(t) for t in x]
```
Comment (blumenstiel) [inline terratorch/models/encoder_decoder_factory.py:39]: I like your approach, thanks for starting this draft!

Other models in terratorch process data in the format [B, C, T, H, W]. That is also the format how data is provided from the generic data modules. It might be good to follow this this pattern und iterate over dim=2 instead of expecting a list.

File: terratorch/models/encoder_decoder_factory.py [hunk 2]
```diff
@@ -20,6 +21,31 @@
...
+
+class TemporalWrapper(nn.Module):
+    def __init__(self, encoder, pooling="mean"):
+        super().__init__()
+        self.encoder = encoder
+        if pooling == "mean":
+            self.pooling = torch.mean
+        elif pooling == "max":
+            self.pooling = torch.max
+        else:
```
Comment (blumenstiel) [inline terratorch/models/encoder_decoder_factory.py:33]: It would be good to have a method `concat` which merges the embeddings of all timestamps along the embedding dim. E.g. for testing how much accuracy we loose if the time stamps are averaged before the decoder.

Comment (blumenstiel) [reply terratorch/models/encoder_decoder_factory.py:33]: This requires fix time stamps defined by the user so that the decoder gets the correct `out_channels`.

**Comment by blumenstiel** [comment]
@rijuld Thanks for working on this feature, that will be very helpful!
To pass the checks, please remember to sign off your next commits (either `git commit -s` or VS Code/PyCharm have some checkboxes for it to do it automatically)
See https://github.com/IBM/terratorch/pull/355/checks?check_run_id=35365530364

**Comment by romeokienzler** [comment]
Dear @rijuld - this has gone stale, any chance to address the conflicts and comments? Then we can merge. You can also join the [TerraTorch community calls](https://github.com/IBM/terratorch/blob/main/CONTRIBUTING.md) or ping us on [Slack](https://join.slack.com/t/terratorch/shared_invite/zt-31tegwss3-t3B4cful2JhcQ02ptGCMJQ)

**Comment by Foxigod** [comment]
Hi, do we have an ETA for this feature/PR merge?

**Comment by Joao-L-S-Almeida** [comment]
@Foxigod I don't know what ETA means in this context. 

@rijuld I tested your implementation and it seems working as expected. For example:
```
def test_temporal_wrapper_prithvi_forward_shapes(dummy_encoder):

    NUM_CHANNELS = 6
    encoder = BACKBONE_REGISTRY.build("prithvi_eo_v2_300")

    wrapper = TemporalWrapper(encoder)
    batch_size = 2

    # Test case 2: Valid input shape
    x = torch.randn(batch_size, NUM_CHANNELS, 4, 224, 224)  # [B, C, T, H, W]
    output = wrapper(x)
    print([o.shape for o in output])

```
It produces 24 tensors with size `[2, 197, 1024]`. It would be nice if you could add some tests for some backbones available on the repository, as `prithvi_eo_v2_300` and `prithvi_swin_B`. 

@paolo-fraccaro Do you think we convention adopted by @rijuld, (B, C,T,H,W), is the most convenient way to use it or it should have a different order, as (B,T,C,H,W) ?

**Comment by romeokienzler** [comment]
@Joao-L-S-Almeida ETA -> Estimated Time of Arrival -> when will it be merged
@rijuld -> can we get some tests as we need to make sure the test coverage doesn't decline

In case u need this urgently we merge today in the hope we still get the tests :)

**Comment by rijuld** [comment]
> @Joao-L-S-Almeida ETA -> Estimated Time of Arrival -> when will it be merged @rijuld -> can we get some tests as we need to make sure the test coverage doesn't decline
> 
> In case u need this urgently we merge today in the hope we still get the tests :)

@romeokienzler Sure, how can I check test coverage while testing?

**Comment by Foxigod** [comment]
@romeokienzler This functionality would be beneficial for our case, and I believe I've heard @paolofraccaro mention the same thing. For my case, I don't need it _today_, but in the next week or two would be best if you want to finalize the tests before merging.

**Comment by naomi-simumba** [comment]
@Joao-L-S-Almeida, for the current use cases we have, switching to a `[B, T, C, H, W]` format would be more convenient.
Also, I have tested this for 4 timestamps with the following backbones (with format switched to  `[B, T, C, H, W]`  and image size 224):

- `prithvi_eo_v2_300`/`prithvi_eo_v2_600`
- `ssl4eos12_resnet50_sentinel2_all_decur`
- `ssl4eos12_resnet50_sentinel2_all_dino`

Behaviour with Swin models needs further testing before merge

**Comment by Joao-L-S-Almeida** [comment]
Hi, @rijuld 
You can use a command like:
`pytest -s --cov=terratorch -v --cov-report term-missing tests` (you need the dependencies under `test` in the `pyproject.toml`) running inside the root `terratorch` directory. 
In any case, it's more important to verify if the backbones are working with the new approach than the coverage itself.

**Comment by Joao-L-S-Almeida** [comment]
@rijuld I added two modifications to your fork as a PR. I need you approve them in order to we finish the first part of this task.

**Comment by rijuld** [comment]
> @rijuld I added two modifications to your fork as a PR. I need you approve them in order to we finish the first part of this task. I think after it we can merge and maybe continue in a new PR after it.

Hi @Joao-L-S-Almeida I have approved the workflow

**Comment by Joao-L-S-Almeida** [comment]
Thanks, @rijuld. The tests passed. Could you merge them ?

**Comment by rijuld** [comment]
> Hi @Joao-L-S-Almeida

Hi @Joao-L-S-Almeida done

**Comment by Joao-L-S-Almeida** [comment]
@rijuld Thank you for it. I'll merge it, since it seems alright for now. Feel free to add any others changes and contributions in new PRs.
